src/spuco/last_layer_retrain/dfr.py

```diff
@@ -82,8 +82,9 @@ def train_single_model(self, C, X_train, y_train, g_train, class_weight):
         :param class_weight: Weight associated with each class.
         :type class_weight: dict or 'balanced', optional
         """
+        group_names = {g for g in g_train}
         group_partition = []
-        for g in range(np.max(g_train)+1):
+        for g in group_names:
```

Comment:
made small change so that it works when some groups missing, e.g., only having groups 0 1 3.